Api changes to support async api

ianoc opened the pull request:
This is the API boundary core, without an actual impl.

The test updates here cover the reply.rs async version.


Future PR's needed:

1) Make integration tests support multiple impls
2) Add async examples (1 per PR)
2) Add tokio impl, and update examples to use it


While these are inflight building with all flags enabled will see some unused code. But normal build flags should see no unused code warnings.

wmanley commented:
> The test updates here cover the reply.rs async version.

I've been working on a sans-io `Reply` type which should help reduce duplication here. I've raised my WIP as #131 to show.

cberner commented on src/lib.rs:
```diff
@@ -40,6 +40,7 @@ use std::cmp::max;
 #[cfg(feature = "abi-7-13")]
 use std::cmp::min;
 
+pub mod async_api;
```
I'm not sure, but I think you can add `#[cfg(feature = "async_impl")]` here, and then you won't need to add it all over the `async_api` module

ianoc replied:
i'll add an async_api flag, but i don't think it'll avoid that since building agnostic crates requires a separation between exposed api vs particular implementation i believe

cberner replied:
Ah, I see. I misunderstood why the `cfg`s were there. Either way works for me, so feel free to change it back in a future PR if you want

cberner commented on src/async_api/reply.rs:
```diff
@@ -0,0 +1,1336 @@
+        nlink: attr.nlink,
+        uid: attr.uid,
+        gid: attr.gid,
+        rdev: attr.rdev,
+        flags: attr.flags,
+        #[cfg(feature = "abi-7-9")]
+        blksize: attr.blksize,
+        #[cfg(feature = "abi-7-9")]
+        padding: attr.padding,
+    }
+}
+
+/// Returns a fuse_attr from FileAttr
+#[cfg(not(target_os = "macos"))]
+fn fuse_attr_from_attr(attr: &FileAttr) -> fuse_attr {
+    let (atime_secs, atime_nanos) = time_from_system_time(&attr.atime);
+    let (mtime_secs, mtime_nanos) = time_from_system_time(&attr.mtime);
+    let (ctime_secs, ctime_nanos) = time_from_system_time(&attr.ctime);
+
+    fuse_attr {
+        ino: attr.ino,
+        size: attr.size,
+        blocks: attr.blocks,
+        atime: atime_secs,
+        mtime: mtime_secs,
+        ctime: ctime_secs,
+        atimensec: atime_nanos,
+        mtimensec: mtime_nanos,
+        ctimensec: ctime_nanos,
+        mode: mode_from_kind_and_perm(attr.kind, attr.perm),
+        nlink: attr.nlink,
+        uid: attr.uid,
+        gid: attr.gid,
+        rdev: attr.rdev,
+        #[cfg(feature = "abi-7-9")]
+        blksize: attr.blksize,
+        #[cfg(feature = "abi-7-9")]
+        padding: attr.padding,
+    }
+}
```
Instead of copying these, can you import them from the sync `reply.rs`?

ianoc replied:
👍

cberner reviewed (approved):
Could you add compiling with this feature flag to the CI, and also add `cargo clippy --all-targets --features async_impl` to the `pre` target of the Makefile? Just want to be sure we have some basic test coverage to be sure build errors don't creep into this code, because it's not compiled by default.

I had some other minor comments too. Otherwise it looks good to me though

cberner commented:
> > The test updates here cover the reply.rs async version.
> 
> I've been working on a sans-io `Reply` type which should help reduce duplication here. I've raised my WIP as #131 to show.

Awesome, that'll be great to cut down on duplication! I'll take a look at that PR next.

ianoc commented on Cargo.toml:
```diff
@@ -23,12 +23,16 @@ users = "0.11.0"
 page_size = "0.4.2"
 serde = {version = "1.0.102", features = ["std", "derive"], optional = true}
 zerocopy = "0.3"
+async-trait = "0.1.42"
+futures = "0.3.13"
+derivative = "2.2.0"
```
I can add another flag for that -- right now these are used based on there being the 'async code' vs async impl. The thought here is you can build a crate foo to depend on fuser with default flags, and implement the actual fs using any impl since the API doesn't change -- the final crate/bin/etc could then add the right async impl flag without a conflict and start it with that.

ianoc commented on src/async_api/active_session.rs:
```diff
@@ -0,0 +1,192 @@
+pub(in crate::async_api) trait Worker: Send + Sync {
+    // this method can presumed to only return when the worker/session is being shut down
+    // generally something like a one shot channel will back this.
+    async fn return_on_destory(&mut self) -> ();
```
sgtm

ianoc commented:
I updated the API function name as you suggested, seems better i think.

I made the dependencies and module optional gated on a `async_api` flag, added new checks around it. and removed the duplication

cberner commented:
Merged. Thanks!
